### PR TITLE
[Serializer] Xml encoder throws exception for valid data

### PR DESCRIPTION
**Pull request by gr1ev0us**
| Q             | A
| ------------- | ---
| Branch?       |  2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21617 
| License       | MIT
| Doc PR        | None

#21617  Xml encoder throws exception for valid data
    
- add tests for bool and object encoding
- fix encoding for object in array and field

### CODE REVIEW

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 1]
```diff
@@ -11,19 +11,29 @@
...
+use Symfony\Component\Serializer\Tests\Fixtures\ScalarDummy;
 
 class XmlEncoderTest extends TestCase
 {
+    /**
+     * @var XmlEncoder
+     */
     private $encoder;
 
+    /**
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: This comment is useless (the IDE will be able to infer the type using the default value).

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 2]
```diff
@@ -524,4 +534,96 @@ protected function getObject()
 
         return $obj;
     }
+
+    /**
+     * @test
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: This block is useless (same below).

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 3]
```diff
@@ -524,4 +534,96 @@ protected function getObject()
...
+     */
+    public function testEncodeXmlWithBoolValue()
+    {
+        $expectedXml = '<?xml version="1.0"?>
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: You may use the heredoc syntax (the IDE will be able to highlight the snippet). (Just suggestion, not mandatory).

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 4]
```diff
@@ -524,4 +534,96 @@ protected function getObject()
...
+<response><foo>1</foo><bar>0</bar></response>
+';
+
+        $actualXml = $this->encoder->encode(array('foo' => true, 'bar' => false), 'xml');
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: I don't get why this is expected? IMO it should be `array('foo' => '1', 'bar' => '0')` unless explicitly defined as bool in a XML schema.

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 5]
```diff
@@ -11,19 +11,29 @@
 
 namespace Symfony\Component\Serializer\Tests\Encoder;
 
+use DateTime;
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: For classes in the global namespace, we don't import them (use `\DateTime` directly).

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php (hunk 4 above)
Comment (gr1ev0us) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: This is an actual behavior of XmlEncode, to cast bool to int. I am not sure that i should fix this non explicit behaviour, so i am just cover that with test.

File: src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php [hunk 6]
```diff
@@ -12,18 +12,24 @@
 namespace Symfony\Component\Serializer\Tests\Encoder;
 
 use PHPUnit\Framework\TestCase;
-use Symfony\Component\Serializer\Tests\Fixtures\Dummy;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Encoder/XmlEncoderTest.php]: please revert any change related to ordering: our policy is to add *new* lines in alpha order, but keep old ones as is - to help for merges between branches.

**Comment by fabpot** [comment]
@dunglas Is this one ok for you?

**Comment by dunglas** [comment]
👍

**Comment by fabpot** [comment]
Thank you @gr1ev0us.

**Pull request closed without merging**